### PR TITLE
feat(node): node runner

### PR DESCRIPTION
**Pull request by alanmarkoTrilitech**
# Context
https://app.asana.com/0/1205770721173533/1206345416213081

# Description

This PR makes jstz-node run automatically after sandbox start is executed.

# Manually testing the PR

cargo run --bin jstz -- sandbox start

### CODE REVIEW

File: crates/jstz_node/src/node_runner.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+/// Endpoint defailts for the `jstz-node`
+const ENDPOINT: (&str, u16) = ("127.0.0.1", 8933);
+
+pub async fn run_node(
+    rollup_node_rpc_addr: String,
+    rollup_node_rpc_port: u16,
+    rollup_endpoint: Option<String>,
+    kernel_file_path: String,
+) -> io::Result<()> {
+    env_logger::init_from_env(Env::default().default_filter_or("info"));
```
Comment (johnyob) [inline crates/jstz_node/src/node_runner.rs:19]: We may want to disable the logger in the sandbox, wdyt?

File: crates/jstz_cli/src/sandbox/daemon.rs [hunk 2]
```diff
@@ -245,8 +245,15 @@ pub fn main(cfg: &mut Config) -> Result<()> {
     println!("Saving sandbox config");
     cfg.save()?;
 
+    // Run node
+    let jstz_node = OctezThread::from_child(
+        Command::new("cargo")
+            .args(["run", "--bin", "jstz-node"])
+            .spawn()?,
+    );
+
```
Comment (johnyob) [inline crates/jstz_cli/src/sandbox/daemon.rs:254]: We shouldn't be using `cargo` like this, since this won't be packaged correctly on the user's installation (only works in our local dev environment)

File: crates/jstz_node/src/main.rs [hunk 3]
```diff
@@ -32,44 +27,18 @@ struct Args {
     #[arg(short, long)]
     rollup_endpoint: Option<String>,
 
-    #[arg(long)]
+    #[arg(long, default_value = DEFAULT_KERNEL_FILE_PATH)]
     kernel_file_path: String,
 }
 
 #[actix_web::main]
```
Comment (johnyob) [inline crates/jstz_node/src/main.rs:34]: To avoid issues in future, we should explicitly switch to `#[tokio::main]` (which `actix` uses internally)

File: crates/jstz_node/src/main.rs [hunk 4]
```diff
@@ -1,26 +1,21 @@
...
+    },
+    node_runner::run_node,
+};
 pub use error::{Error, Result};
 
+mod config;
 mod error;
+mod node_runner;
 mod services;
 mod tailed_file;
```
Comment (johnyob) [inline crates/jstz_node/src/main.rs:17]: I think these can be dropped and you simply import them from the crate defined in `jstz_node` (introducing `lib.rs` now defines a crate + a binary in `jstz_node`)

File: crates/jstz_cli/src/sandbox/daemon.rs (hunk 2 above)
Comment (alanmarkoTrilitech) [reply crates/jstz_cli/src/sandbox/daemon.rs:254]: yes, we need to compile it and then change to just 'jstz-node'. For now however, I think it can stay like this and we can create a separate task for changing it.

Comment (johnyob) [reply crates/jstz_cli/src/sandbox/daemon.rs:254]: I'd be keen on tracking it in this PR since the other changes in this PR would be mostly unnecessary if this was our desired approach

File: crates/jstz_node/src/node_runner.rs (hunk 1 above)
Comment (alanmarkoTrilitech) [reply crates/jstz_node/src/node_runner.rs:19]: I've added an option to disable it.

What is the actual purpose of it?

File: crates/jstz_node/src/main.rs (hunk 4 above)
Comment (alanmarkoTrilitech) [reply crates/jstz_node/src/main.rs:17]: we need both crate and binary since we can still run just jstz-node, or run it from the sandbox

File: crates/jstz_cli/src/sandbox/daemon.rs [hunk 5]
```diff
@@ -254,6 +257,26 @@ pub fn main(cfg: &mut Config) -> Result<()> {
...
+                println!("Jstz node started 🎉");
+                let _ = run_node(
+                    DEFAULT_ROLLUP_NODE_RPC_ADDR.to_string(),
+                    DEFAULT_ROLLUP_RPC_PORT,
+                    None,
+                    DEFAULT_KERNEL_FILE_PATH.to_string(),
+                    false,
+                )
+                .await;
+            })
```
[the remark this answers is not in the document]
Comment (alanmarkoTrilitech) [reply crates/jstz_cli/src/sandbox/daemon.rs:274]: This is part of another PR.

File: crates/jstz_cli/src/sandbox/daemon.rs [hunk 6]
```diff
@@ -254,6 +257,26 @@ pub fn main(cfg: &mut Config) -> Result<()> {
...
+            .await
+            .expect("Local task panicked");
+        })
+        .await;
```
[the remark this answers is not in the document]
Comment (alanmarkoTrilitech) [reply crates/jstz_cli/src/sandbox/daemon.rs:278]: Jstz node apparently cannot be moved between threads because it involves some non-thread safe behaviour, so  LocalSet is essentially creating a context for it so that it can be run.

**Comment by johnyob** [comment]
Added a review commit to avoid back-and-forth (feel free to revert if you disagree with the changes):

TL;DR:
- Improved crate/bin separation between `jstz-node` and `jstz_node`
- Re-introduced a cli flag (seems to have been dropped somewhere) for defining the jstz node endpoint of the `jstz-node` 
- Factored out running jstz-node logic into `run_jstz_node` in the sandbox

File: crates/jstz_cli/src/sandbox/daemon.rs [hunk 7]
```diff
@@ -17,9 +13,14 @@ use std::{
     time::Duration,
 };
 use tempfile::TempDir;
+use tokio::task;
 
 use crate::config::{Config, SandboxConfig, SANDBOX_OCTEZ_SMART_ROLLUP_PORT};
 
+const SANDBOX_JSTZ_NODE_ADDR: &str = "127.0.0.1";
+const SANDBOX_JSTZ_NODE_PORT: u16 = 8933;
+const SANDBOX_OCTEZ_SMART_ROLLUP_ADDR: &str = "127.0.0.1";
```
Comment (alanmarkoTrilitech) [inline crates/jstz_cli/src/sandbox/daemon.rs:22]: Now we have it defined it two places which is not ideal. Ideally there should be a shared library for these constants.

File: crates/jstz_node/Cargo.toml [hunk 8]
```diff
@@ -24,7 +24,7 @@ parking_lot = "0.12.1"
 reqwest = { version = "0.11.22", features = ["json"] }
 serde = { version = "1.0.183", features = ["derive"] }
 thiserror = "1.0.56"
-tokio = { version = "1.33.0", features = ["fs"] }
+tokio = { version = "1.33.0", features = ["fs", "rt-multi-thread"] }
```
Comment (alanmarkoTrilitech) [inline crates/jstz_node/Cargo.toml:27]: what does this do?

**Review by alanmarkoTrilitech — COMMENTED** [review]
The changes look good to me, except those constants.

You will have to approve it though.

File: crates/jstz_cli/src/sandbox/daemon.rs (hunk 7 above)
Comment (johnyob) [reply crates/jstz_cli/src/sandbox/daemon.rs:22]: They don't necessarily have to be the same -- the node has defaults for it's CLI and we have defaults for the sandbox. 

They're only currently the same because the node was designed with the sandbox in mind

File: crates/jstz_node/Cargo.toml (hunk 8 above)
Comment (johnyob) [reply crates/jstz_node/Cargo.toml:27]: Tokio seemed to give me a compiler error without this feature flag

**Review by johnyob — APPROVED** [review]
Please squash before merging (especially removing the 'review' commit)
